From ca228bf3e3a4ae6c9921c7295b065f38f30701ca Mon Sep 17 00:00:00 2001
From: Steve Loughran <stevel@apache.org>
Date: Sat, 23 Jul 2016 17:58:46 +0100
Subject: [PATCH 1900/2748] HADOOP-13389 TestS3ATemporaryCredentials.testSTS
 error when using IAM credentials. Contributed by
 Steven K. Wong

(cherry picked from commit 7c46303de3d0d5f9375df0d426121a3885069620)

Change-Id: I8a77d1ed774675aa7bcc09d03858b4333eb54874
---
 .../hadoop/fs/s3a/TestS3ATemporaryCredentials.java |   24 +++++++++-----------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ATemporaryCredentials.java b/hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ATemporaryCredentials.java
index 1ba024d..303f06c 100644
--- a/hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ATemporaryCredentials.java
+++ b/hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ATemporaryCredentials.java
@@ -19,10 +19,10 @@
 package org.apache.hadoop.fs.s3a;
 
 import java.io.IOException;
+import java.net.URI;
 
 import com.amazonaws.auth.AWSCredentials;
-import com.amazonaws.auth.AWSCredentialsProviderChain;
-import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.auth.AWSCredentialsProvider;
 import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClient;
 import com.amazonaws.services.securitytoken.model.GetSessionTokenRequest;
 import com.amazonaws.services.securitytoken.model.GetSessionTokenResult;
@@ -31,6 +31,7 @@
 import org.apache.hadoop.fs.contract.AbstractFSContract;
 import org.apache.hadoop.fs.contract.AbstractFSContractTestBase;
 import org.apache.hadoop.fs.contract.s3a.S3AContract;
+import org.apache.hadoop.fs.s3native.S3xLoginHelper;
 import org.apache.hadoop.conf.Configuration;
 
 import org.junit.Test;
@@ -54,9 +55,6 @@
   private static final Logger LOG =
       LoggerFactory.getLogger(TestS3ATemporaryCredentials.class);
 
-  private S3AFileSystem fs;
-
-
   private static final String PROVIDER_CLASS =
       "org.apache.hadoop.fs.s3a.TemporaryAWSCredentialsProvider";
 
@@ -84,15 +82,15 @@ public void testSTS() throws IOException {
       skip("STS functional tests disabled");
     }
 
-    String parentAccessKey = conf.getTrimmed(ACCESS_KEY, null);
-    String parentSecretKey = conf.getTrimmed(SECRET_KEY, null);
-    String stsEndpoint = conf.getTrimmed(TEST_STS_ENDPOINT, "");
-    AWSCredentialsProviderChain parentCredentials;
-    parentCredentials = new AWSCredentialsProviderChain(
-      new BasicAWSCredentialsProvider(parentAccessKey, parentSecretKey),
-      new InstanceProfileCredentialsProvider()
-    );
+    S3xLoginHelper.Login login = S3AUtils.getAWSAccessKeys(
+        URI.create("s3a://foobar"), conf);
+    if (!login.hasLogin()) {
+      skip("testSTS disabled because AWS credentials not configured");
+    }
+    AWSCredentialsProvider parentCredentials = new BasicAWSCredentialsProvider(
+        login.getUser(), login.getPassword());
 
+    String stsEndpoint = conf.getTrimmed(TEST_STS_ENDPOINT, "");
     AWSSecurityTokenServiceClient stsClient;
     stsClient = new AWSSecurityTokenServiceClient(parentCredentials);
     if (!stsEndpoint.isEmpty()) {
-- 
1.7.9.5

